src/callbacks_stage/positivity_adaptive_filter_dg1d.jl

```diff
@@ -0,0 +1,171 @@
+# By default, Julia/LLVM does not use fused multiply-add operations (FMAs).
+# Since these FMAs can increase the performance of many numerical algorithms,
+# we need to opt-in explicitly.
+# See https://ranocha.de/blog/Optimizing_EC_Trixi for further details.
+@muladd begin
+
+
+"""
+
+Adaptive filter for DGMulti to ensure the well-posedness of the entropy
+projection step.
+"""
+# TODO: only support 1D GaussSBP. Multidimension and general DGMulti in future PRs.
+function adaptive_filter!(u, filter,
+                          mesh::DGMultiMesh{1},
+                          equations::Union{AbstractCompressibleEulerEquations,CompressibleNavierStokesDiffusion2D},
+                          dg::DGMulti{1}, cache)
+
+  rd = dg.basis
+  @unpack Vq = rd
+  @unpack invVDM = cache
+  @unpack u_values, u_modal_coeffs, entropy_var_values = cache
+  # @unpack interp_matrix_lobatto_to_gauss, interp_matrix_gauss_to_face = cache
+
+  # TODO: redundant operations with local_filtered_entropy_projection!
+  apply_to_each_field(mul_by!(Vq), u_values, u)
+  apply_to_each_field(mul_by!(invVDM), u_modal_coeffs, u)
+
+  # TODO: redundant operations with local_filtered_entropy_projection!
+  # TODO: only need the last entropy variable to determine the local bound
+  # transform quadrature values to entropy variables
+  cons2entropy!(entropy_var_values, u_values, equations)
+
+  @threaded for e in eachelement(mesh, dg, cache)
```

Comment:
I am thinking about storing the filtering factors \theta in the cache. It is not necessary for adaptive filter since we only need to modify u in stage_limiter!, but for operator blending (scaled extrapolation operator), \theta are needed to be stored in cache in order for rhs! to appropriately change the extrapolation operator. Maybe we can care about this later?

Reply:
I think we should treat operator scaling separately from filtering, so I wouldn't worry about that now. 

Is there any chance we could store the filtering factors in the filtering struct itself? Modifying the cache would mean we'd have to add the limiting factors to the cache for each type of `DGMulti` solver.

Reply:
Sure. But since AdaptiveFilter is a stage_limiter!, can rhs! and entropy_projection! still have the access to the AdaptiveFilter object? Seems to me only the ODE solver will have the access to stage_limiter! ?

Reply:
Hm, that is a good point...let me think about that. Does `entropy_projection!` currently have access to `AdaptiveFilter` in this PR?

Reply:
No. I guess if you want to access to `AdaptiveFilter` in entropy_projection!, you either pass in as an argument or store it in cache?

Reply:
Actually, I don't think we necessarily need `entropy_projection!` to be aware of the limiter. We filter the conservative variables to enforce bounds after each stage. We could enforce bounds involving the entropy projection, so that when `entropy_projection!` is called, the result should already be bounds-satisfying.

This way, the filter would have the solver, equations, and mesh passed in, but the solver wouldn't need to know about the filter.

Reply:
Yes, unnecessary for the case of `AdaptiveFilter`, and the approach you described is the approach I am taking now. But `rhs!` will need the filter factors when using the operator (extrapolation) blending approach. I will leave the issue for future PRs.

Reply:
That's ok - the operator blending is being dealt with in a separate PR.